### PR TITLE
[Compiler-v2][VM] add metadata check for script

### PR DESCRIPTION
**Pull request by rahxephon89**
## Description

This PR adds a check in the VM when running bytecode generated by an unstable compiler or containing unstable language feature on mainnet

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Added an e2e move test

## Key Areas to Review

1. logic of checking in the VM;
2. coverage in e2e move tests.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 1]
```diff
@@ -709,6 +710,27 @@ impl AptosVM {
 
         let func = session.load_script(script.code(), script.ty_args())?;
 
+        // Check that unstable bytecode cannot be executed on mainnet
+        if self
+            .features()
+            .is_enabled(FeatureFlag::REJECT_UNSTABLE_BYTECODE_FOR_SCRIPT)
```
Comment (rahxephon89) [inline aptos-move/aptos-vm/src/aptos_vm.rs:716]: Do we need to check both `REJECT_UNSTABLE_BYTECODE_FOR_SCRIPT` and `REJECT_UNSTABLE_BYTECODE` given that `REJECT_UNSTABLE_BYTECODE` is already enabled?

Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:716]: I think a new flag only is enough

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 2]
```diff
@@ -709,6 +710,27 @@ impl AptosVM {
...
+        {
+            let script = match CompiledScript::deserialize_with_config(
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/aptos_vm.rs:718]: Can we cache the deserialization of compiled scripts here by bundling it with `verify_no_event_emission_in_script` (this also deserializes the script) under a single function which does the deserialization and then runs verification passes?

Comment (rahxephon89) [reply aptos-move/aptos-vm/src/aptos_vm.rs:734]: done

Comment (georgemitenkov) [reply aptos-move/aptos-vm/src/aptos_vm.rs:734]: Why else branch in the code below? I think having
```
// deserialize here
if feature_enabled {
  self.reject_unstable_bytecode_for_script(&script)?;
}
verifier::event_validation::verify_no_event_emission_in_compiled_script(&script)?;
```
is sufficient?

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 3]
```diff
@@ -1610,6 +1634,22 @@ impl AptosVM {
...
+    /// Check whether the script can be run on mainnet based on the unstable tag in the metadata
+    pub fn reject_unstable_bytecode_for_script(&self, module: &CompiledScript) -> VMResult<()> {
+        if self.chain_id().is_mainnet() {
+            if let Some(metadata) =
+                aptos_framework::get_compilation_metadata_from_compiled_script(module)
+            {
+                if metadata.unstable {
+                    return Err(PartialVMError::new(StatusCode::UNSTABLE_BYTECODE_REJECTED)
+                        .with_message("script marked unstable cannot be run on mainnet".to_string())
+                        .finish(Location::Undefined));
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/aptos_vm.rs:1646]: nit: probably can use `Location::Script`

**Review by georgemitenkov — APPROVED** [review]
Looks good, only cone comment about if-else: I think you can just deserialise outside of if branch and then use a single if statement to make things nicer.

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 2 above)
Comment (rahxephon89) [reply aptos-move/aptos-vm/src/aptos_vm.rs]: done

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 3 above)
Comment (rahxephon89) [reply aptos-move/aptos-vm/src/aptos_vm.rs]: done

**Review by georgemitenkov — APPROVED** [review]

**Review by ziaptos — APPROVED** [review]
good

**Pull request merged**